internal/pkg/template/templates/app/cf.yml

```
@@ -82,6 +82,16 @@ Resources:
               AWS:
                 - !Sub arn:${AWS::Partition}:iam::${AWS::AccountId}:root{{range $accounts}}
                 - !Sub arn:${AWS::Partition}:iam::{{.}}:root{{end}}
+          - Action: s3:PutObject
+            Effect: Deny
+            Resource: !Sub arn:${AWS::Partition}:s3:::${PipelineBuiltArtifactBucket}/*
+            Principal: '*'
+            Condition:
+              StringNotEqualsIfExists:
+                s3:x-amz-server-side-encryption: 'aws:kms'
+              'Null':
+                s3:x-amz-server-side-encryption: false
```

```suggestion
            Condition:
              StringNotEquals:
                s3:x-amz-server-side-encryption: 'aws:kms'
              'Null':
                s3:x-amz-server-side-encryption: true
```
? i _think_ (based on [this](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html#Conditions_Null)) that the `false` condition would actually allow unencrypted uploads if the `s3:x-amz-server-side-encryption` header was not present 🤔 (this blog post uses `true`! https://aws.amazon.com/blogs/security/how-to-prevent-uploads-of-unencrypted-objects-to-amazon-s3/)

---

Hmm 🤔 
I think the example you linked uses `true` to say something like: "Deny PutObject when the server-side-encryption header is _not_ present". However, when it is not present, then S3 uses the default encryption setting, which is KMS in our case (as expected). In the tutorial you linked, the default encryption of S3 is not used, as far as I can see, right? Maybe that's why they added such a policy?
Changing the condition to true would require every client to always provide the server-side-encryption header in the request, isn't it?
And combining the `Null` and `StringNotEquals` condition would effectively only be true (denying the action) when the condition key is not set, isn't it? Still allowing values other than `aws:kms` (as the null condition will not be true anymore and conditions are AND combined?) 🤔

The current policy says something like this:
If the action is PutObject
And the condition key `s3:x-amz-server-side-encryption` is not null
And the condition key `s3:x-amz-server-side-encryption` is not `aws:kms`
Then deny the action.

If the `s3:x-amz-server-side-encryption` key is null, the policy _should not_ deny the action, as the buckets default encryption (aws:kms) would be used in this case.

Which should be exactly what we want, isn't it? 🤔 I tried the above policy in the policy simulator, just to make sure that it seems to work as expected. I used the following policy:
```
{
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "s3:*",
            "Resource": "*"
        },
        {
            "Effect": "Deny",
            "Action": "s3:PutObject",
            "Resource": "*",
            "Condition": {
                "StringNotEqualsIfExists": {
                    "s3:x-amz-server-side-encryption": "aws:kms"
                },
                "Null": {
                    "s3:x-amz-server-side-encryption": false
                }
            }
        }
    ]
}
```

And then using the `PutObject` action with `*` as a resource. Then different variations of the `s3:x-amz-server-side-encryption` condition value (empty, `aws:kms` and `AES256`). For me it seems to work as expected?

---

> I think the example you linked uses true to say something like: "Deny PutObject when the server-side-encryption header is not present".

This is my understanding as well! Deny `PutObject` if the header is not present 👍 

Since we agree on the _meaning_ of the policy, it seems the outstanding question is: Do we need to require the header or not? If we do, then we should use `StringNotEquals` and `Null: true`; If we don't want to require the header we should use `StringNotEqualsIfExists` and `Null: false`.

Based on [the CloudFormation docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-serversideencryptionbydefault.html), I think you're right! Specifically,
> if a PUT Object request doesn't specify any server-side encryption, this default encryption will be applied.

where "default encryption" is the value specified in the CFN (`aws:kms` as part of this PR).

---

All this to say - looks good to me! Thanks again for making this change!